### PR TITLE
MTDSA-13941 Prepare API Common Code for TYS Changes

### PR DESCRIPTION
**Pull request by sean-flynn-HMRC**

### CODE REVIEW

File: app/config/FeatureSwitches.scala [hunk 1]
```diff
@@ -18,24 +18,32 @@ package config
...
-      config    <- value
-      enabled   <- config.getOptional[Boolean](s"version-$versionNo.enabled")
+      validVersion <- maybeVersion
+      enabled      <- featureSwitchConfig.getOptional[Boolean](s"version-$validVersion.enabled")
     } yield enabled
 
     enabled.getOrElse(false)
   }
 
+  val isV1R7cRoutingEnabled: Boolean        = isEnabled("v1r7c-endpoints.enabled")
```
Comment (harry-ritchie) [inline app/config/FeatureSwitches.scala:40]: Why do we need `isV1R7cRoutingEnabled` and `isTaxYearNotEndedRuleEnabled` here? They weren't in the code before so wondering what the reason is for needing them?

File: app/v1/connectors/IfsUri.scala [hunk 2]
```diff
@@ -16,4 +16,4 @@
 
 package v1.connectors
 
-case class IfsUri[Resp](value: String)
+//case class IfsUri[Resp](value: String)
```
Comment (harry-ritchie) [inline app/v1/connectors/IfsUri.scala:19]: Maybe we could just delete this file instead if we do not need it?

File: app/v1/connectors/httpparsers/StandardIfsHttpParser.scala [hunk 3]
```diff
@@ -16,25 +16,26 @@
...
 import play.api.http.Status._
 import play.api.libs.json.Reads
 import uk.gov.hmrc.http.{HttpReads, HttpResponse}
-import v1.connectors.IfsOutcome
-import v1.models.errors.{DownstreamError, OutboundError}
+import v1.connectors.DownstreamOutcome
+import v1.models.errors.DownstreamError
 import v1.models.outcomes.ResponseWrapper
 
 object StandardIfsHttpParser extends HttpParser {
```
Comment (harry-ritchie) [inline app/v1/connectors/httpparsers/StandardIfsHttpParser.scala:27]: Could rename the file (+ object) to `StandardDownsteamHttpParser` ?

File: app/v1/models/domain/TaxYear.scala [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  /** The tax year in the pre-TYS downstream format, e.g. "2024".
+    */
+  val asDownstream: String = value
+
+  /** The tax year in the Tax Year Specific downstream format, e.g. "23-24".
+    */
+  val asTysDownstream: String = {
+    val year2 = value.toInt - 2000
+    val year1 = year2 - 1
+    s"${year1}-$year2"
```
Comment (harry-ritchie) [inline app/v1/models/domain/TaxYear.scala:50]: Don't need the {} for `year1`

File: app/v1/models/domain/TaxYear.scala [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+  /** Use this for downstream API endpoints that are known to be TYS.
+    */
+  def useTaxYearSpecificApi(implicit featureSwitches: FeatureSwitches): Boolean = featureSwitches.isTaxYearSpecificApiEnabled && year >= 2024
+
+  override def toString: String = s"TaxYear($value)"
+}
+
+object TaxYear {
```
Comment (harry-ritchie) [inline app/v1/models/domain/TaxYear.scala:60]: Could also include the helper methods (see [here](https://github.com/hmrc/self-assessment-accounts-api/pull/276/files) in TaxYear). If so, will also need to update TaxYearSpec

File: app/v1/models/errors/downstreamErrors.scala [hunk 6]
```diff
@@ -14,24 +14,25 @@
  * limitations under the License.
  */
```
Comment (harry-ritchie) [inline app/v1/models/errors/downstreamErrors.scala:16]: Is it better to rename this file to `DownstreamErrors`? i.e. not lowercase d

File: app/v1/support/IfsResponseMappingSupport.scala [hunk 7]
```diff
@@ -24,7 +25,7 @@ import v1.models.outcomes.ResponseWrapper
 trait IfsResponseMappingSupport {
```
Comment (harry-ritchie) [inline app/v1/support/IfsResponseMappingSupport.scala:25]: As you've changed the generics to `DownstreamError`, should this class be specific to IFS? If not, better to change all references to IFS to `Downstream`

File: test/mocks/MockAppConfig.scala [hunk 8]
```diff
@@ -40,7 +40,7 @@ trait MockAppConfig extends MockFactory {
     def minimumPermittedTaxYear: CallHandler[Int] = (mockAppConfig.minimumPermittedTaxYear _).expects()
```
Comment (harry-ritchie) [inline test/mocks/MockAppConfig.scala:41]: I think we also need TYS IFS Config in this file

File: test/v1/connectors/BaseDownstreamConnectorSpec.scala [hunk 9]
```diff
@@ -35,7 +36,7 @@ class BaseDownstreamConnectorSpec extends ConnectorSpec {
   val url         = "some/url?param=value"
   val absoluteUrl = s"$baseUrl/$url"
 
-  implicit val httpReads: HttpReads[IfsOutcome[Result]] = mock[HttpReads[IfsOutcome[Result]]]
+  implicit val httpReads: HttpReads[DownstreamOutcome[Result]] = mock[HttpReads[DownstreamOutcome[Result]]]
 
   class Test(ifsEnvironmentHeaders: Option[Seq[String]]) extends MockHttpClient with MockAppConfig {
```
Comment (harry-ritchie) [inline test/v1/connectors/BaseDownstreamConnectorSpec.scala:41]: Think we also need to include TysIfs tests.

File: test/v1/connectors/httpparsers/StandardIfsHttpParserSpec.scala [hunk 10]
```diff
@@ -40,7 +41,7 @@ class StandardIfsHttpParserSpec extends UnitSpec {
```
Comment (harry-ritchie) [inline test/v1/connectors/httpparsers/StandardIfsHttpParserSpec.scala:41]: Will also need to rename this file to `Downstream...` if you change the api file

File: test/v1/support/IfsResponseMappingSupportSpec.scala [hunk 11]
```diff
@@ -47,14 +48,14 @@ class IfsResponseMappingSupportSpec extends UnitSpec {
     "single error" when {
       "the error code is in the map provided" must {
         "use the mapping and wrap" in {
-          mapping.mapIfsErrors(errorCodeMap)(ResponseWrapper(correlationId, IfsErrors.single(IfsErrorCode("ERR1")))) shouldBe
+          mapping.mapIfsErrors(errorCodeMap)(ResponseWrapper(correlationId, DownstreamErrors.single(DownstreamErrorCode("ERR1")))) shouldBe
             ErrorWrapper(correlationId, Error1)
```
Comment (harry-ritchie) [inline test/v1/support/IfsResponseMappingSupportSpec.scala:52]: Again, maybe change file name to `DownstreamResponseMappingSupportSpec` and if so rename references to IFS to Downstream?

File: test/v1/connectors/BaseDownstreamConnectorSpec.scala (hunk 9 above)
Comment (harry-ritchie) [reply test/v1/connectors/BaseDownstreamConnectorSpec.scala:41]: Will also need to update the `ConnectorSpec` file with the TYS config

File: app/v1/models/domain/TaxYear.scala (hunk 5 above)
Comment (sean-flynn-HMRC) [reply app/v1/models/domain/TaxYear.scala:60]: I'm not sure it's best practice to add functionality if it's not going to be used. Perhaps we should leave the taxYear object as it is and add each helper method as we need them (if we ever need them in this API)?

File: app/v1/support/IfsResponseMappingSupport.scala (hunk 7 above)
Comment (sean-flynn-HMRC) [reply app/v1/support/IfsResponseMappingSupport.scala:25]: Yes

File: test/v1/connectors/ConnectorSpec.scala [hunk 12]
```diff
@@ -42,6 +46,13 @@ trait ConnectorSpec extends UnitSpec with Status with MimeTypes with HeaderNames
       Some("other-deductions-api")
     )
 
+  val dummyHeaderCarrierConfig: HeaderCarrier.Config =
+    HeaderCarrier.Config(
+      Seq("^not-test-BaseUrl?$".r),
+      Seq.empty[String],
+      Some("individuals-income-received-api")
```
Comment (harry-ritchie) [inline test/v1/connectors/ConnectorSpec.scala:53]: Think`Some("individuals-income-received-api")` should be `Some("other-deductions-api")`

Comment (harry-ritchie) [reply test/v1/connectors/ConnectorSpec.scala:53]: And in that case, could either rename it to `dummyTysIfsHeaderCarrierConfig` or just delete `dummyIfsHeaderCarrierConfig ` and keep `dummyHeaderCarrierConfig` as a generic one (As it is correct for both TYS and non-TYS)

**Review by harry-ritchie — APPROVED** [review]

**Pull request merged**